beacon/light/committee_chain.go

```diff
@@ -0,0 +1,555 @@
+			log.Crit("Inconsistent database error: first committee is not in the fixed roots range")
+		}
+		if s.committees.AfterLast > s.fixedRoots.AfterLast && s.committees.AfterLast > s.updates.AfterLast+1 {
+			log.Crit("Inconsistent database error: last committee is neither in the fixed roots range nor proven by updates")
+		}
+		log.Trace("Sync committee chain loaded", "first period", s.committees.First, "last period", s.committees.AfterLast-1)
+	}
+	// roll back invalid updates (might be necessary if forks have been changed since last time)
+	var batch ethdb.Batch
+	for !s.updates.IsEmpty() {
+		if update := s.updates.get(s.updates.AfterLast - 1); update == nil || s.verifyUpdate(update) {
+			if update == nil {
+				log.Crit("Sync committee update missing", "period", s.updates.AfterLast-1)
+			}
+			break
+		}
+		if batch == nil {
+			batch = s.db.NewBatch()
+		}
+		s.rollback(batch, s.updates.AfterLast)
+	}
+	if batch != nil {
+		if err := batch.Write(); err != nil {
+			log.Error("Error writing batch into chain database", "error", err)
+		}
+	}
+	return s
+}
+
+// Reset resets the committee chain.
+func (s *CommitteeChain) Reset() {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	batch := s.db.NewBatch()
+	s.rollback(batch, 0)
+	if err := batch.Write(); err != nil {
+		log.Error("Error writing batch into chain database", "error", err)
+	}
+}
+
+// AddFixedRoot sets a fixed committee root at the given period.
+// Note that the period where the first committee is added has to have a fixed
+// root which can either come from a CheckpointData or a trusted source.
+func (s *CommitteeChain) AddFixedRoot(period uint64, root common.Hash) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	batch := s.db.NewBatch()
+	oldRoot := s.getCommitteeRoot(period)
+	if !s.fixedRoots.CanExpand(period) {
+		if root != oldRoot {
+			return ErrInvalidPeriod
+		}
+		for p := s.fixedRoots.AfterLast; p <= period; p++ {
+			s.fixedRoots.add(batch, p, s.getCommitteeRoot(p))
+		}
+	}
+	if oldRoot != (common.Hash{}) && (oldRoot != root) {
+		// existing old root was different, we have to reorg the chain
+		s.rollback(batch, period)
+	}
+	s.fixedRoots.add(batch, period, root)
+	if err := batch.Write(); err != nil {
+		log.Error("Error writing batch into chain database", "error", err)
+		return err
+	}
+	return nil
+}
+
+// DeleteFixedRootsFrom deletes fixed roots starting from the given period.
+// It also maintains chain consistency, meaning that it also deletes updates and
+// committees if they are no longer supported by a valid update chain.
+func (s *CommitteeChain) DeleteFixedRootsFrom(period uint64) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if period >= s.fixedRoots.AfterLast {
+		return nil
+	}
+	batch := s.db.NewBatch()
+	s.fixedRoots.deleteFrom(batch, period)
+	if s.updates.IsEmpty() || period <= s.updates.First {
+		s.updates.deleteFrom(batch, period)
+		s.deleteCommitteesFrom(batch, period)
+	} else {
+		fromPeriod := s.updates.AfterLast + 1
```

Comment:
If `period` is larger than `s.update.First` but less then `s.update.AfterLast`, shouldn't we also delete a part of updates?

Instead, we keep all the updates and just remove the committee after the `s.updates.AfterLast  +1`. Won't it result in a gap between fixedRoot and updates?

Reply:
Looks correct to me, added some explanatory comments